cmd/producer/functions.go

```diff
@@ -336,3 +372,56 @@ func optionalMap(s string, items []string) []int {
 	// Parse provided percentages
 	return mustMap(s)
 }
+
+func mustProducerMode(s string) producerMode {
+	v := strings.ToLower(strings.Trim(os.Getenv(s), " "))
+	switch v {
+	case "stdout":
+		return modeStdout
+	case "http":
+		return modeHTTP
+	case "http2":
+		return modeHTTP2
+	case "pulsar":
+		return modePulsar
+	default:
+		panic(fmt.Errorf("producer mode out of the known domain: %s", s))
+	}
+}
+
+func newProducer(slotName string, mode producerMode, useOneClientPerSlot bool) (publisherCloser, error) {
+	switch mode {
+	case modeHTTP:
+		return producer.NewHTTPProducer(slotName, os.Environ())
+	case modeHTTP2:
+		return producer.NewHTTP2Producer(slotName, os.Environ())
+	case modeStdout:
+		return producer.NewStdoutPublisher(slotName), nil
+	case modePulsar:
+		if !useOneClientPerSlot {
+			return nil, fmt.Errorf("pulsar mode requires useOneClientPerSlot to be true")
+		}
+		return producer.NewPulsarProducer(slotName, os.Environ())
+	default:
+		return nil, fmt.Errorf("unknown mode: %s", mode)
+	}
+}
+
+func getHostname(hostname string) (string, int, error) {
+	match := hostnameRE.FindStringSubmatch(hostname)
+	if len(match) <= 2 {
+		return "", 0, fmt.Errorf("hostname is invalid: %s", hostname)
+	}
+
+	instanceNumber, err := strconv.Atoi(match[1])
+	if err != nil {
+		return "", 0, fmt.Errorf("error getting instance number from hostname %s: %v", hostname, err)
+	}
+
+	deploymentName := match[2]
+	if deploymentName == "" {
+		return "", 0, fmt.Errorf("deployment name is empty: %s", hostname)
+	}
```

Comment:
Can we add some tests here as the hostname is an issue encountered in the past? Sorry for the duplicate :)